File: ATHLON/IDENTITYSERVICE/src/main/java/com/athlon/identityservice/organization/entity/AcademyFinance.java
package com.athlon.identityservice.organization.entity;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;

@Entity
@Table(name = "academy_finances")
public class AcademyFinance {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "finance_id", updatable = false, nullable = false)
    private Long financeId;

    @Column(name = "finance_uuid", updatable = false, nullable = false, unique = true)
    private UUID financeUuid;

    @Column(name = "organization_id", nullable = false)
    private Long organizationId;

    @Column(name = "organization_uuid", nullable = false)
    private UUID organizationUuid;

    @Column(name = "transaction_type", nullable = false, length = 20)
    private String transactionType; // EXPENSE or INCOME

    @Column(name = "category", nullable = false, length = 60)
    private String category; // Tuition Fee, Admission, Coach Salary, Equipment, Facility Rent, Maintenance, Tournament, etc.

    @Column(name = "title", nullable = false, length = 150)
    private String title;

    @Column(name = "amount", nullable = false, precision = 12, scale = 2)
    private BigDecimal amount;

    @Column(name = "transaction_date", nullable = false)
    private LocalDate transactionDate;

    @Column(name = "payment_method", length = 30)
    private String paymentMethod; // UPI, CASH, BANK_TRANSFER, CARD, CHEQUE

    @Column(name = "paid_to_or_by", length = 100)
    private String paidToOrBy; // Student name, Coach name, or Vendor name

    @Column(name = "student_uuid")
    private UUID studentUuid;

    @Column(name = "student_name", length = 150)
    private String studentName;

    @Column(name = "batch_uuid")
    private UUID batchUuid;

    @Column(name = "batch_name", length = 100)
    private String batchName;

    @Column(name = "invoice_number", length = 50)
    private String invoiceNumber;

    @Column(name = "fee_status", length = 30)
    private String feeStatus; // PAID, PARTIAL, PENDING, REFUNDED

    @Column(name = "notes", length = 500)
    private String notes;

    @Column(name = "receipt_url", length = 255)
    private String receiptUrl;

    @Column(name = "created_by")
    private Long createdBy;

    @Column(name = "updated_by")
    private Long updatedBy;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public AcademyFinance() {
    }

    @PrePersist
    protected void onCreate() {
        if (this.financeUuid == null) {
            this.financeUuid = UUID.randomUUID();
        }
        if (this.transactionDate == null) {
            this.transactionDate = LocalDate.now();
        }
        if (this.feeStatus == null) {
            this.feeStatus = "PAID";
        }
        if (this.createdAt == null) {
            this.createdAt = LocalDateTime.now();
        }
        this.updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    public Long getFinanceId() {
        return financeId;
    }

    public void setFinanceId(Long financeId) {
        this.financeId = financeId;
    }

    public UUID getFinanceUuid() {
        return financeUuid;
    }

    public void setFinanceUuid(UUID financeUuid) {
        this.financeUuid = financeUuid;
    }

    public Long getOrganizationId() {
        return organizationId;
    }

    public void setOrganizationId(Long organizationId) {
        this.organizationId = organizationId;
    }

    public UUID getOrganizationUuid() {
        return organizationUuid;
    }

    public void setOrganizationUuid(UUID organizationUuid) {
        this.organizationUuid = organizationUuid;
    }

    public String getTransactionType() {
        return transactionType;
    }

    public void setTransactionType(String transactionType) {
        this.transactionType = transactionType;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public BigDecimal getAmount() {
        return amount;
    }

    public void setAmount(BigDecimal amount) {
        this.amount = amount;
    }

    public LocalDate getTransactionDate() {
        return transactionDate;
    }

    public void setTransactionDate(LocalDate transactionDate) {
        this.transactionDate = transactionDate;
    }

    public String getPaymentMethod() {
        return paymentMethod;
    }

    public void setPaymentMethod(String paymentMethod) {
        this.paymentMethod = paymentMethod;
    }

    public String getPaidToOrBy() {
        return paidToOrBy;
    }

    public void setPaidToOrBy(String paidToOrBy) {
        this.paidToOrBy = paidToOrBy;
    }

    public UUID getStudentUuid() {
        return studentUuid;
    }

    public void setStudentUuid(UUID studentUuid) {
        this.studentUuid = studentUuid;
    }

    public String getStudentName() {
        return studentName;
    }

    public void setStudentName(String studentName) {
        this.studentName = studentName;
    }

    public UUID getBatchUuid() {
        return batchUuid;
    }

    public void setBatchUuid(UUID batchUuid) {
        this.batchUuid = batchUuid;
    }

    public String getBatchName() {
        return batchName;
    }

    public void setBatchName(String batchName) {
        this.batchName = batchName;
    }

    public String getInvoiceNumber() {
        return invoiceNumber;
    }

    public void setInvoiceNumber(String invoiceNumber) {
        this.invoiceNumber = invoiceNumber;
    }

    public String getFeeStatus() {
        return feeStatus;
    }

    public void setFeeStatus(String feeStatus) {
        this.feeStatus = feeStatus;
    }

    public String getNotes() {
        return notes;
    }

    public void setNotes(String notes) {
        this.notes = notes;
    }

    public String getReceiptUrl() {
        return receiptUrl;
    }

    public void setReceiptUrl(String receiptUrl) {
        this.receiptUrl = receiptUrl;
    }

    public Long getCreatedBy() {
        return createdBy;
    }

    public void setCreatedBy(Long createdBy) {
        this.createdBy = createdBy;
    }

    public Long getUpdatedBy() {
        return updatedBy;
    }

    public void setUpdatedBy(Long updatedBy) {
        this.updatedBy = updatedBy;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }
}

File: athlon-user/src/lib/api/academyFinance.ts
import { api } from './client';

export interface AcademyFinance {
  financeId?: number;
  financeUuid?: string;
  organizationId?: number;
  organizationUuid: string;
  transactionType: 'EXPENSE' | 'INCOME';
  category: string;
  title: string;
  amount: number;
  transactionDate: string;
  paymentMethod?: string;
  paidToOrBy?: string;
  studentUuid?: string;
  studentName?: string;
  batchUuid?: string;
  batchName?: string;
  invoiceNumber?: string;
  feeStatus?: 'PAID' | 'PARTIAL' | 'PENDING' | 'REFUNDED';
  notes?: string;
  receiptUrl?: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface AcademyFinanceSummary {
  totalIncome: number;
  totalExpense: number;
  netBalance: number;
  totalFeeCollected: number;
  pendingFeeAmount: number;
  transactionCount: number;
  studentFeePaymentCount: number;
  expenseByCategory: Record<string, number>;
  incomeByCategory: Record<string, number>;
}

export interface CreateAcademyFinancePayload {
  organizationUuid: string;
  transactionType: 'EXPENSE' | 'INCOME';
  category: string;
  title: string;
  amount: number;
  transactionDate: string;
  paymentMethod?: string;
  paidToOrBy?: string;
  studentUuid?: string;
  studentName?: string;
  batchUuid?: string;
  batchName?: string;
  invoiceNumber?: string;
  feeStatus?: 'PAID' | 'PARTIAL' | 'PENDING' | 'REFUNDED';
  notes?: string;
  receiptUrl?: string;
}

export interface UpdateAcademyFinancePayload {
  financeUuid: string;
  transactionType?: 'EXPENSE' | 'INCOME';
  category?: string;
  title?: string;
  amount?: number;
  transactionDate?: string;
  paymentMethod?: string;
  paidToOrBy?: string;
  studentUuid?: string;
  studentName?: string;
  batchUuid?: string;
  batchName?: string;
  invoiceNumber?: string;
  feeStatus?: 'PAID' | 'PARTIAL' | 'PENDING' | 'REFUNDED';
  notes?: string;
  receiptUrl?: string;
}

export const AcademyFinanceService = {
  getFinances: async (
    orgUuid: string,
    type?: string,
    category?: string,
    studentUuid?: string,
    batchUuid?: string,
    startDate?: string,
    endDate?: string
  ): Promise<AcademyFinance[]> => {
    let url = `/api/identity/academy/finances/org/${orgUuid}?`;
    if (type && type !== 'ALL') url += `type=${type}&`;
    if (category && category !== 'ALL') url += `category=${encodeURIComponent(category)}&`;
    if (studentUuid) url += `studentUuid=${studentUuid}&`;
    if (batchUuid && batchUuid !== 'ALL') url += `batchUuid=${batchUuid}&`;
    if (startDate) url += `startDate=${startDate}&`;
    if (endDate) url += `endDate=${endDate}&`;

    const res = await api.get<any>(url);
    if (Array.isArray(res)) return res;
    if (res && Array.isArray(res.data)) return res.data;
    return [];
  },

  createFinance: async (payload: CreateAcademyFinancePayload): Promise<AcademyFinance> => {
    const res = await api.post<any>('/api/identity/academy/finances/add', payload);
    return (res && res.data) ? res.data : res;
  },

  updateFinance: async (payload: UpdateAcademyFinancePayload): Promise<AcademyFinance> => {
    const res = await api.post<any>('/api/identity/academy/finances/update', payload);
    return (res && res.data) ? res.data : res;
  },

  deleteFinance: (financeUuid: string) =>
    api.post<void>(`/api/identity/academy/finances/delete/${financeUuid}`, {}),

  getSummary: async (orgUuid: string, startDate?: string, endDate?: string): Promise<AcademyFinanceSummary | null> => {
    let url = `/api/identity/academy/finances/summary/org/${orgUuid}?`;
    if (startDate) url += `startDate=${startDate}&`;
    if (endDate) url += `endDate=${endDate}&`;

    const res = await api.get<any>(url);
    if (res && res.data) return res.data;
    return res || null;
  }
};
